File: app/src/main/java/com/example/yogurtmachine/ui/YogurtMachineViewModel.java
package com.example.yogurtmachine.ui;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.example.yogurtmachine.data.YogurtMachineState;
import com.example.yogurtmachine.network.MqttService;

public class YogurtMachineViewModel extends ViewModel {
    private final MqttService mqttService;
    private final MutableLiveData<YogurtMachineState> uiState;

    public YogurtMachineViewModel() {
        mqttService = new MqttService();
        uiState = new MutableLiveData<>(new YogurtMachineState());
        
        // Observe MQTT service state changes
        mqttService.getMachineState().observeForever(state -> {
            if (state != null) {
                uiState.setValue(state);
            }
        });
        
        connectToMqtt();
    }

    private void connectToMqtt() {
        mqttService.connect();
    }

    public LiveData<YogurtMachineState> getUiState() {
        return uiState;
    }

    public MqttService getMqttService() {
        return mqttService;
    }

    public void startProcess(float targetTemp, int timeHours) {
        mqttService.startProcess(targetTemp, timeHours);
    }

    public void stopProcess() {
        mqttService.stopProcess();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        mqttService.disconnect();
    }
} 

File: app/src/main/java/com/example/yogurtmachine/ui/YogurtMachineViewModel.kt
package com.example.yogurtmachine.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.yogurtmachine.data.YogurtMachineState
import com.example.yogurtmachine.network.MqttService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class YogurtMachineViewModel : ViewModel() {
    private val mqttService = MqttService()
    
    private val _uiState = MutableStateFlow(YogurtMachineState())
    val uiState: StateFlow<YogurtMachineState> = _uiState.asStateFlow()

    init {
        viewModelScope.launch {
            mqttService.machineState.collect { state ->
                _uiState.value = state
            }
        }
        connectToMqtt()
    }

    private fun connectToMqtt() {
        mqttService.connect()
    }

    fun startProcess(targetTemp: Float, timeHours: Int) {
        mqttService.startProcess(targetTemp, timeHours)
    }

    fun stopProcess() {
        mqttService.stopProcess()
    }

    override fun onCleared() {
        super.onCleared()
        mqttService.disconnect()
    }
} 